### ISSUE TITLE
calculation results off by 1

### ISSUE THREAD

**Issue by dsnz** [fast-reflexes/BirthdayProblem-Python#1]
Hi,

many calculations give a result 1 less than the real one
eg.

D= 1,000,000,000, P= 50%  =>  N= 32,733  by Taylor (correct  32,734)
D= 83, P= 50%  =>  N= 11  by Taylor (correct  12)
D= 69, P= 50%  =>  N= 10  by Taylor (correct  11)
etc. by comparing with https://oeis.org/A033810/list

it's true that many examples in internet have the same issue (eg. SAS blog)
however for such a dedicated project an absolutely correct result is expected

NOTE:
1. I have run the examples using the web-site and not the code of this repo directly so there may be issues
related to interfacing the code with the web-site
2. for D such that sqrt(D) <= 50,000 it's ok to use the normal definition formula with multiplication of about sqrt(D) factors,
it's quite fast in python and the correct result compared to any other approximation formula

**Comment by fast-reflexes** [comment]
Hi there! Thanks a lot for filing this bug report! I agree that it must be fixed. I have a lot on my table right now but I will take a look at it as soon as I can. I will get back to you if I have any more questions or when I have corrected the issue. Thanks a lot! 🙏

**Comment by fast-reflexes** [comment]
@dsnz I've looked into this a bit and it doesn't seem to be an issue connected to numerical instability but rather that the Taylor calculations are an approximation = not exact. At least the calculations yield the same results forward and backwards (e.g. P(D, N) and N(P, D) yield results (from Taylor approximation) that align.

My only solution to this currently would be to add terms to the Taylor polynomial to make it more exact, or are you thinking along some other lines?

Given the formula `n(P, d) = sqrt(2 * d * ln(1 / P))` (which can be found on the Wikipedia page for example), for d = 69 and P = 0.5, we arrive at `sqrt(2 * 69 * ln(1 / 0.5)) = sqrt(2 * 69 * ln(2))` which is approx. 9.78 which ceiled becomes 10 (which is the wrong answer because 11 is the right answer) but that's the calculation I'm offering for calulcations of `n(P, d)`... I mean I would LOVE to make it more accurate but since it's not a problem with numerical stability (which I first thought), the method itself must be refined for the results to be more accurate and I'm neither sure of the direction to take here nor if that direction would be reasonable :)

**Comment by fast-reflexes** [comment]
Interesting ... or maybe not so interesting but using the approximation formula

`P(n, d) = 1 - e^(-(n^2 / 2d))` (`P1`) which is more of an approximation than `P(n, d) = 1 - e^(-((n * (n-1) / 2d))` (`P2`)

We can see that `P1(10, 69) ~ 0.51` but `P2(10, 69) ~ 0.48`, thus with P1, 10 is the correct answer but with P2 it is not. The use of `n^2` instead of `n * (n - 1)` then causes a deviation from the right answer (no surprise there) so in the reverse formula for `n(P, d) = sqrt(-ln(1 - P) * 2d)`, could we solve `n(n-1) = -ln(1 - P) * 2d` instead of `n^2 = -ln(1 - P) * 2d`, the result would be more accurate. I think this aspect is more pertinent to pursue than increasing the terms in the Taylor polynomial ( if that's even possible for this approximation).

The best way to achieve this however is probably to use the approximation formula with `n^2` and `Px` and then test the result numerically in the forward calculation `P(n, d)` (with `n * (n - 1)`) using both `n(Px, d)` and `n(Px, d) + 1` and see which of these yield a result `P(n, d) >= Px`.

**Comment by dsnz** [comment]
Hi,

you researched it well indeed

and yes using **n * (n-1)** will definitely result in more correct results
(either by solving for n or doing the nice trick you suggested of calculating **both n(Px, d) and n(Px, d) +1**
and see which results in P(n, d) >= Px)

I can only add that **any formula is an approximation based on inequalities**

and I would suggest that for values of d such that the formula

**n(P, d) = sqrt(2 * d * ln(1 / (1-P)))**  (and this is the correct formula, using 1-P instead of P)

will result in a value of less than 100,000 for example then
then use the exact formula **1 - (1 - 1/d) . (1 - 2/d) . .. . (1 - (n-1)/d)** of about 100,000 multiplications
until **such n** that the result is **>= P**
ie. "**brute force the resul**t", since it's very fast anyway on most processors

**Comment by fast-reflexes** [comment]
> **n(P, d) = sqrt(2 * d * ln(1 / (1-P)))** (and this is the correct formula, using 1-P instead of P)

Yes but `ln(1 / (1 - P) = ln(1) - ln(1 - P) = 0 - ln(1 - P) = -ln(1 - P)` which is what I'm using. Or I didn't understand what you meant by the "correct" formula? `ln(1 / ( 1 - P)) = -ln(1 - P)` so that's correct and from the `n^2` vs `n * (n - 1)` discussion, the above formula is the less exact since it uses `sqrt` which is derived from the fact that we have `n^2` while we would really like to have `n * (n - 1)` instead :)

I think the best path is to do both the numerical trick in the forward calculation and what you suggest. As long as it takes less than 2 seconds I'm fine with it. I have attempted to implement the trick and it works out nicely. To be continued :)

**Comment by dsnz** [comment]
Hi,

about ln(1-P) , I just saw a ln(P) in your first post, may be a typo because of fast typing .. I'm sure the correct formula is used in code

I'm also sure you will improve the result along the lines we discussed :)
and expect to read your final report and code improvements

it seems ridiculous to put so much effort for the most accurate result (just -1 offset) for such big numbers :D
and which accurate result has no practical benefit 
but it's necessary for authoritative code and this is the gist of this issue

**Comment by fast-reflexes** [comment]
@dsnz Fixed this now. I added an exact method to the calculation of `n(P, d)` as well. I refrained from implementing my previously mentioned heuristic since I want to keep the Taylor approximation clean, e.g. it IS an approximation and it remains an approximation honoring its definition at all times. The rationale is that for relatively small values where +-1 matters, the exact solution will give the correct result and when the exact solution is infeasible to use, +-1 might not matter = the Taylor approximation works fine as it is. Also, the approximation will work better for high numbers too so this should also limit the risk of the +-1 error margin.

Let me know what you think and if you have anything to add, otherwise I will close this ticket and thanks a lot for contributing and driving this project forward by giving comments and suggestions :) 🎉
